Chore/documentation (#490)
cozy/cozy-client: docs/react-integration.md

## What is the Cozy-Client React Integration?

In addition to the Javascript client, Cozy-Client provides a way to connect directly the data from your Cozy instance to your React components. 

Once connected, your components will receive the requesting data and a fetch status in their props. The data will automatically be refreshed upon modification.

## Setup

The following procedure requires you to already know how to initialize a CozyClient instance and create a query to fetch documents.

### 1. Initialize a CozyClient provider

> Load `CozyClient` and `CozyProvider` objects

Reviewer: ```suggestion
Import `CozyClient` and `CozyProvider`
```

```javascript
import CozyClient, { CozyProvider } from 'cozy-client'
```

> Initialize a CozyClient instance (see the `CozyClient()` documentation for additionnal parameters, like to provide a schema)

Reviewer: ```suggestion
Initialize a CozyClient instance (see the `CozyClient()` documentation for additional parameters, for example to provide a schema)
```

```javascript
const client = new CozyClient({
  uri: 'http://cozy.tools:8080',
  token: '...'
})
```

> Then wrap you React application inside a the CozyProvider component with your newly instance of CozyClient.

Reviewer: ```suggestion
Then wrap you React application inside a CozyProvider instance with your newly instance of CozyClient.
```

  applyMiddleware(myMiddleware)
)

function MyCozyApp(props) {
  return <CozyProvider client={client} store={store}>
    <MyApp {...props} />
  </CozyProvider>
}

ReactDOM.render(<MyCozyApp />,
  document.getElementById('main')
)
```

### 2.a Requesting data with `<Query />`

To make it easy to fetch data and make it available to your component, we provide a Render Props component called `Query`. Basic example of usage:

```jsx
import { Query } from 'cozy-client'

const todos = 'io.cozy.todos'
const checked = { checked: false }
const query = client => client.find(todos).where(checked)

function TodoList() {
  return <Query query={query}>
    {
      ({ data, fetchStatus }) =>
      fetchStatus !== 'loaded'
        ? <h1>Loading...</h1>
        : <ul>{data.map(todo => <li>{todo.label}</li>)}</ul>
    }
  </Query>
}
```

When we use `Query` to "wrap" a component, three things happen:
 - The query passed as a prop will be executed when `Query` mounts, resulting in the loading of data from the client-side store, or the server if the data is not in the store;
 - `Query` subscribes to the store, so that it is updated if the data changes;

>  - `Query` pass the result of the query as props to the children function.

Reviewer: ```suggestion
 - `Query` passes the result of the query as props to the children function.
```

The following props will be given to your wrapped component:
 - `data`: an array of documents
 - `fetchStatus`: the status of the fetch (`pending`, `loading`, `loaded` or `error`)
 - `lastFetch`: when the last fetch occured
 - `hasMore`: the fetches being paginated, this property indicates if there are more documents to load

You can also pass a function intead of a direct query. Your function will be given the props of the component and should return the requested query:

```jsx
import { Query } from 'cozy-client'

const query = function (props) {
  const todos = 'io.cozy.todos'
  const where = { checked: props.checked }
  return client => client.find(todos).where(where)
}

function TodoList() {

>   return <Query query={query}>

Reviewer: It think its for the use with `queryConnect` ?

```
queryConnect({
  checkedTodos: {
    as: 'checkedTodos',
    query: props => client => client.find('io.cozy.todos').where({ checked: props.checked })
  }
})
```

>   return <Query query={query}>

Reviewer 2: Yes it seems.

>   return <Query query={query}>

Author: I indeed missed the `checked={props.checked}` from `<Query />` . This form allows to externalize the query, and have it constant on each render.

Anyways, I am documenting what does exists. If we think it should not exist, then let's remove the corresponding code.

`queryConnect`, through `withQuery`, has it's own mapping of `queryOpts = typeof queryOpts === 'function' ? queryOpts(props) : queryOpts`.

    {
      ({ data, fetchStatus }) =>
      fetchStatus !== 'loaded'
        ? <h1>Loading...</h1>
        : <ul>{data.map(todo => <li>{todo.label}</li>)}</ul>
    }
  </Query>
}

// use <TodoList checked={false} />
```

Note: Your query will be binded when the `<Query/>` component mounts. Future changes in props will not modify the query.

### 2.b Requesting data with the `withQuery` HOC

At your preference, you could use an high order component. `withQuery` will take the name of the props field where it should send the result of the query and the actual query:

```jsx
import { withQuery } from 'cozy-client'

function TodoList(props) {
  const { data, fetchStatus } = props.result
  if (fetchStatus !== 'loaded') {
    return <h1>Loading...</h1>
  } else {
    return <ul>{data.map(todo => <li>{todo.label}</li>)}</ul>
  }
}

const dest = 'result'

const todos = 'io.cozy.todos'
const checked = { checked: false }
const query = (client, props) => client.find('id', props.id).where(checked)

const ConnectedTodoList = withQuery(dest, query)(TodoList)
```

`withQuery` will use `<Query />` internally so you will inherits the same behaviour.

> If you want to connect multiple queries you could also use `queryConnect` which takes all destination fields and queries as keyed object:

Author: Ok for me. Let me update the doc

```jsx
import { queryConnect } from 'cozy-client'

function TodoList(props) {
  const { data, fetchStatus } = props.checked
  if (fetchStatus !== 'loaded') {
    return <h1>Loading...</h1>
  } else {
    return <ul>{data.map(todo => <li>{todo.label}</li>)}</ul>
  }
}

const todos = 'io.cozy.todos'
const checked = client => client.find(todos).where({ checked: false })
const archived = client => client.find(todos).where({ archive: false })

const queries = { checked, archived }

const ConnectedTodoList = queryConnect(queries)(TodoList)
```

### 3. Mutating data

The simplier way is to use the `withClient` high order component. It will inject a `client` in your props with the CozyClient instance you gave to the `<CozyProvider />` upper.

```javascript
import { queryConnect } from 'cozy-client'

function TodoList(props) {
  const { client } = props
  // use client.create()
  return <div />
}

const ConnectedTodoList = withClient(TodoList)
```

> Alternatively, you have a `withMutation()` which returns HOC. This HOC forwards you a `createDocument`, a `saveDocument` and a `deleteDocument` in your props. They are the `create()`, `save()`and `delete()` functions from CozyClient, binded to the instance you gave to the `<CozyProvider />`.

Reviewer 2: ```suggestion
Alternatively, you have a `withMutation()` which returns HOC. This HOC forwards you a `createDocument`, a `saveDocument` and a `deleteDocument` in your props. They are the `create()`, `save()`and `delete()` functions from CozyClient, bound to the instance you gave to the `<CozyProvider />`.
```

```javascript

> import { withMutation } from 'cozy-client'

Reviewer 2: ```suggestion
import { withMutations } from 'cozy-client'
```

function TodoList(props) {

>   const { client } = props

Reviewer 2: ```suggestion
  const { createDocument } = props
```

>   // use client.create()

Reviewer 2: We should show a full example here too.

  return <div />
}

const ConnectedTodoList = withMutations()(TodoList)
```

> Finally, `<Query />` also take a `mutations` optional props. It should have a function that will receive the CozyClient instance, the query requested and the rest of props given to the component, and should return a keyed object which will be added to the props of your wrapped component.

Reviewer 2: ```suggestion
Finally, `<Query />` also takes a `mutations` optional props. It should have a function that will receive the CozyClient instance, the query requested and the rest of props given to the component, and should return a keyed object which will be added to the props of your wrapped component.
```
